README.md

```diff
@@ -1,8 +1,132 @@
+> flooded.”
+>
+> ― Lemony Snicket, Horseradish
+
+Bidirectional TOML parser and renderer. See the following blog post with more details:
+
+* TODO: insert link to blog post
+
+This README contains basic usage example of `tomland`. All code below can be
+compiled and run with the following command:
+
+```
+cabal new-build && cabal new-exec readme
+```
+
+## Preamble: imports and language extensions
+
+Since this is a literate haskell file, we need to specify all our language
+extensions and imports up front.
+
+```haskell
+{-# OPTIONS -Wno-unused-top-binds #-}
+
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes       #-}
+
+import Control.Applicative ((<|>))
+import Control.Category ((>>>))
+import Data.Text (Text)
+import Toml (TomlBiMap, TomlCodec, (.=))
+
+import qualified Data.Text.IO as TIO
+import qualified NeatInterpolation as Interpol
+import qualified Toml
+```
+
+## Data type: parsing and printing
+
+Consider the following TOML configuration specified as interpolated text via
+`neat-interpolation` quasi-quoter:
+
+```haskell
+tomlExample :: Text
+tomlExample = [Interpol.text|
+server.port = 8080
+server.description = "This is production server"
+
+[mail]
+    host = "smtp.gmail.com"
+    send-if-inactive = false
+|]
+```
+
+This static configuration is captured by the following Haskell data type:
+
+```haskell
+data Settings = Settings
+    { settingsPort        :: !Port
+    , settingsDescription :: !Text
+    , settingsMail        :: !Mail
+    , settingsUsers       :: ![User]
+    }
+
+data Mail = Mail
+    { mailHost           :: !Host
+    , mailSendIfInactive :: !Bool
+    }
+
+data User
+    = Admin  Integer  -- id of admin
+    | Client Text     -- name of the client
+    deriving (Show)
+
+newtype Port = Port Int
+newtype Host = Host Text
+```
+
+Using `tomland` library you can write bidirectional converters for these types
+in the following way:
+
+```haskell
+settingsCodec :: TomlCodec Settings
+settingsCodec = Settings
+    <$> Toml.diwrap (Toml.int "server.port")       .= settingsPort
+    <*> Toml.text             "server.description" .= settingsDescription
+    <*> Toml.table mailCodec  "mail"               .= settingsMail
+    <*> pure []                                    .= settingsUsers
+--    <*> Toml.list  userCodec  "user"               .= settingsUsers
+
+mailCodec :: TomlCodec Mail
+mailCodec = Mail
+    <$> Toml.diwrap (Toml.text "host") .= mailHost
+    <*> Toml.bool "send-if-inactive"   .= mailSendIfInactive
+
+_Admin :: TomlBiMap User Integer
+_Admin = Toml.prism Admin $ \case
+    Admin i -> Right i
+    other   -> Toml.wrongConstructor "Admin" other
+
+_Client :: TomlBiMap User Text
+_Client = Toml.prism Client $ \case
+    Client n -> Right n
+    other    -> Toml.wrongConstructor "Client" other
+
+userCodec :: TomlCodec User
+userCodec =
+        Toml.match (_Admin >>> Toml._Integer) "id"
+    <|> Toml.match (_Client >>> Toml._Text) "name"
+```
+
```

Comment:
Probably add some comment before the code about some functions that would be used in here, like `diwrap`, `match`. More about which is suitable for which kind of data types: sum types, type aliases etc. What do you think?

Reply:
Nice idea! This indeed will be very helpful to understand the code :slightly_smiling_face: